lightning/src/routing/router.rs

```diff
@@ -876,40 +878,126 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
+				first_hop_targets.get(&first_hop_in_route.src_node_id).is_some() ||
+				network.get_nodes().get(&first_hop_in_route.src_node_id).is_some();
 			if have_hop_src_in_graph {
-				// BOLT 11 doesn't allow inclusion of features for the last hop hints, which
-				// really sucks, cause we're gonna need that eventually.
-				let last_hop_htlc_minimum_msat: u64 = match hop.htlc_minimum_msat {
-					Some(htlc_minimum_msat) => htlc_minimum_msat,
-					None => 0
-				};
-				let directional_info = DummyDirectionalChannelInfo {
-					cltv_expiry_delta: hop.cltv_expiry_delta as u32,
-					htlc_minimum_msat: last_hop_htlc_minimum_msat,
-					htlc_maximum_msat: hop.htlc_maximum_msat,
-					fees: hop.fees,
-				};
-				// We assume that the recipient only included route hints for routes which had
-				// sufficient value to route `final_value_msat`. Note that in the case of "0-value"
-				// invoices where the invoice does not specify value this may not be the case, but
-				// better to include the hints than not.
-				if add_entry!(hop.short_channel_id, hop.src_node_id, payee, directional_info, Some((final_value_msat + 999) / 1000), &empty_channel_features, 0, path_value_msat, 0) {
-					// If this hop connects to a node with which we have a direct channel,
-					// ignore the network graph and, if the last hop was added, add our
-					// direct channel to the candidate set.
-					//
-					// Note that we *must* check if the last hop was added as `add_entry`
-					// always assumes that the third argument is a node to which we have a
-					// path.
-					if let Some(&(ref first_hop, ref features, ref outbound_capacity_msat, _)) = first_hop_targets.get(&hop.src_node_id) {
-						add_entry!(first_hop, *our_node_id , hop.src_node_id, dummy_directional_info, Some(outbound_capacity_msat / 1000), features, 0, path_value_msat, 0);
+				// We start building the path from reverse, i.e., from payee 
+				// to the first RouteHintHop in the path.
+				let route_it = route.iter().rev();
+				let mut prev_route_it = route.iter().rev();
+				let mut aggregate_next_hops_fee_msat = 0;
+				let mut aggregate_next_hops_path_htlc_minimum_msat = 0;
+				let mut hop_used = true;
+				let mut prev_hop;
+
+				for (idx, hop) in route_it.enumerate(){
+					// BOLT 11 doesn't allow inclusion of features for the last hop hints, which
+					// really sucks, cause we're gonna need that eventually.
+					let hop_htlc_minimum_msat: u64 = match hop.htlc_minimum_msat {
+						Some(htlc_minimum_msat) => htlc_minimum_msat,
+						None => 0
+					};
+
+					let directional_info = DummyDirectionalChannelInfo {
+						cltv_expiry_delta: hop.cltv_expiry_delta as u32,
+						htlc_minimum_msat: hop_htlc_minimum_msat,
+						htlc_maximum_msat: hop.htlc_maximum_msat,
+						fees: hop.fees,
+					};
+
+					prev_hop = match idx {
+						0 => &payee,
+						_ => match prev_route_it.next() {
+							Some(x) => &x.src_node_id,
+							None => &payee,
+						}
+					};
+
+					if idx == 0 {
```

Comment:
Because aggregate_next_hops_fee_msat needs to be zero at the last RouteHintHop which connects to the payee. Since we are building a path from payee then the first value of aggregate_next_hops_fee_msat has to be 0 since no fees will be charged for routing after that hop.